Zend/zend_operators.h

```diff
@@ -49,7 +49,7 @@
 #include "zend_multiply.h"
 #include "zend_object_handlers.h"
 
-#define LONG_SIGN_MASK ZEND_LONG_MIN
+#define LONG_SIGN_MASK (((zend_ulong)1) << (8*sizeof(zend_long)-1))
```

Comment:
The spec is complicated, but I'm not sure this is actually undefined behavior. FWIU, `&` is well defined for both signed and unsigned integers. In the case of two integers with the same conversion rank, the signed integer will be converted to an unsigned one.

In any case, I find the new and explicit solution a bit more obvious.

Reply:
The old code `(((zend_long)1) << (8*sizeof(zend_long)-1))` (note zend_ulong->zend_long next to the `1`) was UB. UBSAN complained at this computation.

Reply:
The old yes, but the new doesn't. `ZEND_LONG_MIN` should be equivalent to only the signed bit being set. The only difference between the old and new solution is the signedness, if I understand correctly. Anyway, this looks fine, this wasn't meant as an argument, just trying to understand. :slightly_smiling_face:

Reply:
Ah right, I guess I mixed up `ZEND_LONG_MIN` with 32/64 bit longs... I'll change that back.

Reply:
In theory, your new solution is better because it supports all complements supported by C. In practice, it likely doesn't matter.

Reply:
Maybe so. However, we rely on 2s complement in a lot of other places, can you even get a 1s complement system somewhere nowadays?

Reply:
Absolutely no clue. :slightly_smiling_face:

Reply:
We could add a respective check to configure.ac.